browser/src/Renderer/WebGLRenderer/TextRenderer/LigatureGrouper/OpenTypeLigatureGrouper.ts

```diff
@@ -0,0 +1,112 @@
+import fontManager from "font-manager"
+import fontkit, { Font } from "fontkit"
+import * as fs from "fs"
+import * as groupBy from "lodash/groupBy"
+import * as Log from "oni-core-logging"
+
+import { GlyphInfo } from "./GlyphInfo"
+import { GlyphSubstitutor } from "./GlyphSubstitutor"
+import { ILigatureGrouper } from "./ILigatureGrouper"
+
+const ligatureFeatures = ["calt", "rclt", "liga", "dlig", "clig"]
+
+export class OpenTypeLigatureGrouper implements ILigatureGrouper {
+    private readonly _font = loadFont(this._fontFamily)
+    private readonly _fontHasLigatures = this._font && checkIfFontHasLigatures(this._font)
+    private readonly _glyphSubstitutor = this._fontHasLigatures
+        ? new GlyphSubstitutor(this._font)
+        : null
+    private readonly _cache = new Map<string, string[]>()
+
+    constructor(private _fontFamily: string) {}
+
+    public getLigatureGroups(characters: string[]) {
+        if (!this._fontHasLigatures) {
+            return characters
+        }
+
+        const concatenatedCharacters = characters.join("")
+
+        const cachedLigatureGroups = this._cache.get(concatenatedCharacters)
+        if (cachedLigatureGroups) {
+            return cachedLigatureGroups
+        }
+
+        const fontGlyphs = this._font.glyphsForString(concatenatedCharacters)
```

Comment:
Note that fontkit's `layout` function does a lot more than just getting glyphs. See https://github.com/foliojs/fontkit/blob/master/src/layout/LayoutEngine.js. You might want to use that instead of doing it yourself manually.

Reply:
I had a thorough look at what all the parts of fontkit do and I tried to replicate and adapt only the really necessary parts. For us, that means only the GSUB replacements were relevant. Other features like kerning, fractional positioning or similar are not relevant in our grid-based editor (It is still using Neovim, after all).